core/embed/rust/src/ui/model_tt/component/hs_render.rs

```diff
@@ -0,0 +1,578 @@
+const HOMESCREEN_IMAGE_SCALE: i16 = 2;
+const HOMESCREEN_IMAGE_SIZE: i16 = WIDTH / HOMESCREEN_IMAGE_SCALE;
+const NOTIFICATION_HEIGHT: i16 = 32;
+const NOTIFICATION_BORDER: i16 = 8;
+const NOTIFICATION_ICON_SPACE: i16 = 8;
+const NOTIFICATION_TEXT_OFFSET: Offset = Offset::new(1, -2);
+const TEXT_ICON_SPACE: i16 = 2;
+
+const HOMESCREEN_DIM_HIEGHT: i16 = 95;
+const HOMESCREEN_DIM_START: i16 = HEIGHT - HOMESCREEN_DIM_HIEGHT;
+const HOMESCREEN_DIM: f32 = 0.63;
+
+const LOCKSCREEN_DIM_HIEGHT: i16 = 95;
+const LOCKSCREEN_DIM_START: i16 = HEIGHT - LOCKSCREEN_DIM_HIEGHT;
+const LOCKSCREEN_DIM: f32 = 0.63;
+const LOCKSCREEN_DIM_BG: f32 = 0.33;
+
+const BLUR_SIZE: usize = 7;
+const BLUR_DIV: u32 =
+    ((65536_f32 * (1_f32 - LOCKSCREEN_DIM_BG)) as u32) / ((BLUR_SIZE * BLUR_SIZE) as u32);
+const DECOMP_LINES: usize = BLUR_SIZE + 1;
+const BLUR_RADIUS: i16 = (BLUR_SIZE / 2) as i16;
+
+const COLORS: usize = 3;
+const RED_IDX: usize = 0;
+const GREEN_IDX: usize = 1;
+const BLUE_IDX: usize = 2;
+
+fn homescreen_get_fg_text(
+    y_tmp: i16,
+    text_info: HomescreenTextInfo,
+    text_buffer: &BufferText,
+    fg_buffer: &mut LineBuffer4Bpp,
+) -> bool {
+    if y_tmp >= text_info.0.y0 && y_tmp < text_info.0.y1 {
+        let y_pos = y_tmp - text_info.0.y0;
+        position_buffer(
+            &mut fg_buffer.buffer,
+            &text_buffer.buffer[(y_pos * WIDTH / 2) as usize..((y_pos + 1) * WIDTH / 2) as usize],
+            4,
+            text_info.0.x0,
+            text_info.1,
+        );
+    }
+
+    y_tmp == (text_info.0.y1 - 1)
+}
+
+fn homescreen_get_fg_icon(
+    y_tmp: i16,
+    text_info: HomescreenTextInfo,
+    icon_data: &[u8],
+    fg_buffer: &mut LineBuffer4Bpp,
+) {
+    if let Some(icon_area) = text_info.3 {
+        let icon_size = icon_area.size();
+        if y_tmp >= icon_area.y0 && y_tmp < icon_area.y1 {
+            let y_pos = y_tmp - icon_area.y0;
+            position_buffer(
+                &mut fg_buffer.buffer,
+                &icon_data
+                    [(y_pos * icon_size.x / 2) as usize..((y_pos + 1) * icon_size.x / 2) as usize],
+                4,
+                icon_area.x0,
+                icon_size.x,
+            );
+        }
+    }
+}
+
+fn homescreen_position_text(
+    text: HomescreenText,
+    buffer: &mut BufferText,
+    icon_buffer: &mut [u8],
+) -> HomescreenTextInfo {
+    let (text, text_style, text_offset, icon) = text;
+
+    let text_width = display::text_width(text, text_style.text_font.into());
+    let font_max_height = display::text_max_height(text_style.text_font.into());
+    let font_baseline = display::text_baseline(text_style.text_font.into());
+    let text_width_clamped = text_width.clamp(0, screen().width());
+
+    let icon_size = if let Some(icon) = icon {
+        let (icon_size, icon_data) = toif_info_ensure(icon, ToifFormat::GrayScaleEH);
+        assert!(icon_size.x < HOMESCREEN_MAX_ICON_SIZE);
+        assert!(icon_size.y < HOMESCREEN_MAX_ICON_SIZE);
```

Comment:
Shouldn't this be `<=`?

Reply:
yep https://github.com/trezor/trezor-firmware/commit/3e3b498047c34005fd7993577ce2c6c849f4390b